man/man5/rlm_policy.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_policy 5 "7 December 2004" "" "FreeRADIUS Module"
.SH NAME
rlm_policy \- FreeRADIUS Module
.SH DESCRIPTION
The \fBrlm_policy\fP module implements a simple "policy" language.
.PP
The policy language implemented by this module is simple, and specific
to RADIUS.  It does not implement variables, arrays, loops, goto's, or
any other feature of a real language.  If those features are needed
for your system, we suggest using \fBrlm_perl\fP.
.PP
What the policy module implements is a simple way to look for
attributes in the request packet (or other places), and to add
attributes to the reply packet (or other places) based on those
decisions.  Where the module shines is that it is significantly more
flexible than the old-style \fBusers\fP file.
.PP
The module has one configuration item:
.IP filename
The file where the policy is stored.

.SH POLICY LANGUAGE
.SS Named policies
The policy is composed of a series of named policies.  The following
example defines a policy named "foo".
.PP
.DS
	policy foo {
.br
		...
.br
	}
.DE
.PP
Policy names MAY NOT be the same as attributes in the dictionary.
Defining a policy with the same name as a dictionary attribute will
cause an error message to be printed, and the policy will not be
loaded.
.PP
When the policy module is listed in a module section like "authorize",
the module calls a policy named "authorize".  The "post-auth",
etc. sections behave the same.  These names cannot be changed.
.PP
.DS
	include "policy.txt"
.DE
.PP
The filename must be in a double-quoted string, and is assumed to be
relative to the location of the current file.  If the filename ends
with a '/', then it is assumed to be a directory, and all files in
that directory will be read.
.PP
.DS
	include "dir/"
.DE
.PP
All file in "dir/" will be read and included into the policy
definition.  Any dot files (".", "..", etc.) will not be included,
however.
.PP
.SS Including multiple files
The main file referred to from the \fIradiusd.conf\fP may include one
or more other files, as in the following example.
.PP
.SS Referencing a named policy
The following example references a named policy
.DS
	foo()
.DE
While the brackets are required, no arguments may be passed.
.PP
.SS Conditions
"if" statements are supported.
.PP
	if (expression) {
.br
		...
.br
	}
.DE
.PP
and "else"
.PP
	if (expression) {
.br
		...
.br
	} else {
.br
		...
.br
	}
.DE
.PP
also, "else if"
.PP
	if (expression) {
.br
		...
.br
	} else if (expression) {
.br
		...
.br
	}
.DE
.PP
.SS Expressions within "if" statements
Always have to have brackets around them.  Sorry.
.PP
The following kinds of expressions may be used, with their meanings.
.IP (attribute-reference)
TRUE if the referenced attribute exists, FALSE otherwise.  See below
for details on attribute references.
.IP (!(expression))
FALSE if the expression returned TRUE, and TRUE if the nested expression
returned FALSE.
.IP "(attribute-reference == value)"
Compares the attribute to the value.  The operators here can be "==",
"!=", "=~", "!~", "<", "<=", ">", and ">=".
.IP "(string1 == string2)"
A special case of the above.  The "string1" is dynamically expanded at
run time, while "string2" is not.  The operators here can be "==",
"!=", "=~",and "!~".  Of these, the most useful is "=~', which lets
you do things like ("%{ldap:query...}" =~ "foo=(.*) ").  The results
of the regular expression match are put into %{1}, and can be used
later.  See "doc/variables.txt" for more information.
.IP "((expression1) || (expression2))"
Short-circuit "or".  If expression1 is TRUE, expression2 is not
evaluated.
.IP "((expression1) && (expression2))"
Short-circuit "and".  If expression1 is FALSE, expression2 is not
evaluated.
.IP Limitations.
The && and || operators have equal precedence. You can't call a
function as a expression.
.PP
.PP
.SS Attribute references
Attribute references are:
.IP Attribute-Name
Refers to an attribute of that name in the Access-Request or
Accounting-Request packet.  May also refer to "server-side"
attributes, which are not documented anywhere.
.IP request:Attribute-Name
An alternate way of referencing an attribute in the request packet.
.PP
.IP reply:Attribute-Name
An attribute in the reply packet
.PP
.IP proxy-request:Attribute-Name
An attribute in the Access-Request or Accounting-Request packet which
will be proxied to the home server.
.PP
.IP proxy-reply:Attribute-Name
An attribute in the Access-Accept or other packet which was received
from a home server.
.PP
.IP control:Attribute-Name
An attribute in the per-request configuration and control attributes.
Also known as "check" attributes (doc/variables.txt).
.PP
.PP
.SS Adding attributes to reply packet (or other location)
	reply .= {
.br
		attribute-name = value
.br
		...
.br
		attribute-name = value
.br
	}
.DE
.PP
The first name can be "request", "reply", "control", "proxy-request",
or "proxy-reply".
.PP
The operator can be
.PP
 .= - appends attributes to end of the list
.PP
 := - replaces existing list with the attributes in the list (bad idea)
.PP
 = - use operators from "attribute = value" to decide what to do. (see "users")
.PP
The block must contain only attributes and values.  Nothing else is permitted.

.SH SECTIONS
.BR authorize
.BR post-auth
.BR pre-proxy
.BR post-proxy
.PP
.SH FILES
.I /etc/raddb/radiusd.conf
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR users (5),
.BR radiusd.conf (5)
.SH AUTHOR
Alan DeKok <aland@ox.org>



doc/ChangeLog
FreeRADIUS 2.0.3 ; $Date$ , urgency=medium
	Feature improvements
	* Updated raddb/certs/ca.cnf with extensions to allow ca.der
	  to be imported as a CA on Symbian and Windows Mobile devices.
	  Closes bug #524
	* Enable multiple matches in "hints" via Fall-Through = Yes.
	  Closes bug #477
	* Added preliminary SQLite driver, contibuted by Apple.
	  Untested, with no sample configuration.  This address bug #470.
	* Updated logging sub-system so that log messages from libfreeradius
	  can go to the log file, and not stdout.
	* Added dictionary.rfc5176
	* EAP module now checks for instance name, and uses that for
	  authentication.  This avoids the need to set Auth-Type when
	  there are multiple instances of the EAP module.
	* Added Module-Return-Code attribute, which contains the value
	  returned by the previous module (ok/fail/update/etc.)

	Bug fixes
	* Corrected typos in rlm_dbm.  Closes bugs #521 and #522.
	* Detail file "listen" sections now work much better.
	* Don't allow old "log_*" to over-ride new format.  Closes bug #525
	* Initialize allocated memory in Oracle SQL driver.  This fixes
	  occasional crashes on some systems.  Closes bug #518
	* Call correct function in rlm_protocol_filter.  This enables the
	  module to build.  Closes bug #512.
	* Added deprecated flag to build for rlm_krb5.  This allows it to
	  run on 64-bit systems.  Closes bug #491
	* Corrected error message when parsing invalid configurations
	  so it doesn't crash.  Closes bug #527
	* Fix handling of timeouts in rlm_ldap that affected 64-bit systems.
	* Handle $INCLUDE's in "instantiate" section.  Closes #528.
	* Format updates to "man" pages from Stephen Gran.

FreeRADIUS 2.0.2 ; Date: 2008/02/14 11:13:48  , urgency=medium
	Feature improvements
	* Added notes on how to debug the server in radiusd.conf
	* Moved all "log_*" in radiusd.conf to log{} section.
	  The old configurations are still accepted, though.
	* Added ca.der target in raddb/certs/Makefile.  This is
	  needed for importing CA certs into Windows.
	* Added ability send raw attributes via "Raw-Attribute = 0x0102..."
	  This is available only debug builds.  It can be used
	  to create invalid packets! Use it with care.
	* Permit "unlang" policies inside of Auth-Type{} sub-sections
	  of the authenticate{} section.  This makes some policies easier
	  to implement.
	* "listen" sections can now have "type = proxy".  This lets you
	  control which IP is used for sending proxied requests.
	* Added note on SSL performance to raddb/certs/README

	Bug fixes
	* Fixed reading of "detail" files.
	* Allow inner EAP tunneled sessions to be proxied.
	* Corrected MySQL schemas
	* syslog now works in log{} section.
	* Corrected typo in raddb/certs/client.cnf
	* Updated raddb/sites-available/proxy-inner-tunnel to
	  permit authentication to work.
	* Ignore zero-length attributes in received packets.
	* Correct memcpy when dealing with unknown attributes.
	* Corrected debugging messages in attr_rewrite.
	* Corrected generation of State attribute in EAP.  This
	  fixes the "failed to remember handler" issues.
	* Fall back to DEFAULT realm if no realm was found.
	  Based on a patch from Vincent Magnin.
	* Updated example raddb/sites-available/proxy-inner-tunnel
	* Corrected behavior of attr_filter to match documentation.
	  This is NOT backwards compatible with previous versions!
	  See "man rlm_attr_filter" for details.

FreeRADIUS 2.0.1 ; Date: 2008/01/22 13:29:37  , urgency=low
	Feature improvements
	* "unlang" has been expanded to do less run-time expansion,
	  and to have better handling of typed data.  See "man unlang"
	  for documentation and new examples.
	
	Bug fixes
	* The 'acct_unique' module has been updated to understand
	  the deprecated (but still used) Client-IP-Address attribute.
	* The EAP-MSCHAPv2 module no longer leaks MS-CHAP2-Success in
	  packets.
	* Fixed crash in rlm_dbm.
	* Fixed parsing of syslog configuration.

FreeRADIUS 2.0.0 ; Date: 2007/11/24 08:33:09 , urgency=low
	Feature improvements
	* Debugging mode is much clearer and easier to read.
        * A new policy language makes many configurations trivial.
	  See "man unlang" for a complete description.
	* Virtual servers are now supported.  This permits clear separation
	  of policies.  See raddb/sites-available/README
	* EAP-TLS (PEAP, EAP-TTLS) and OpenSSL certificates "just work".
	  See raddb/certs/README for details.
	* Proxying is much more configurable than before.
	  See proxy.conf for documentation on pools, and new config items.
	* Full support for IPv6.
	* Much more complete support for the RADIUS SNMP MIBs.
	* HUP now works.  Only some modules are re-loaded,
	  and the server configuation is *not* reloaded.
	* "check config" option now works.  See "man radiusd"
	* radrelay functionality is now included in the server core.
	  See raddb/sites-available/copy-acct-to-home-server
	* VMPS support.  It is minimal, but functional.
	* Cleaned up internal API's and names, including library names.
	
	Bug fixes
	* Many.


src/modules/rlm_policy/rlm_policy.c
/*
 * rlm_policy.c		Implements a policy language
 *
 * Version:	$Id$
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Copyright 2004  Alan DeKok <aland@ox.org>
 * Copyright 2006  The FreeRADIUS server project
 */

#include <freeradius-devel/ident.h>
RCSID("$Id$")

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/modules.h>

#include "rlm_policy.h"

/*
 *	A mapping of configuration file names to internal variables.
 *
 *	Note that the string is dynamically allocated, so it MUST
 *	be freed.  When the configuration file parse re-reads the string,
 *	it free's the old one, and strdup's the new one, placing the pointer
 *	to the strdup'd string into 'config.string'.  This gets around
 *	buffer over-flows.
 */
static const CONF_PARSER module_config[] = {
  { "filename",  PW_TYPE_FILENAME,
    offsetof(rlm_policy_t,filename), NULL,  "${confdir}/policy.txt"},

  { NULL, -1, 0, NULL, NULL }		/* end the list */
};


/*
 *	Callbacks for red-black trees.
 */
static int policyname_cmp(const void *a, const void *b)
{
	return strcmp(((const policy_named_t *)a)->name,
		      ((const policy_named_t *)b)->name);
}


/*
 *	Detach a policy.
 */
static int policy_detach(void *instance)
{
	rlm_policy_t *inst = instance;

	if (inst->policies) rbtree_free(inst->policies);
	free(instance);
	return 0;
}

/*
 *	Do any per-module initialization that is separate to each
 *	configured instance of the module.  e.g. set up connections
 *	to external databases, read configuration files, set up
 *	dictionary entries, etc.
 *
 *	If configuration information is given in the config section
 *	that must be referenced in later calls, store a handle to it
 *	in *instance otherwise put a null pointer there.
 */
static int policy_instantiate(CONF_SECTION *conf, void **instance)
{
	rlm_policy_t *inst;

	/*
	 *	Set up a storage area for instance data
	 */
	inst = rad_malloc(sizeof(*inst));
	if (!inst) {
		return -1;
	}
	memset(inst, 0, sizeof(*inst));

	/*
	 *	If the configuration parameters can't be parsed, then
	 *	fail.
	 */
	if (cf_section_parse(conf, inst, module_config) < 0) {
		policy_detach(inst);
		return -1;
	}

	inst->policies = rbtree_create(policyname_cmp, rlm_policy_free_item, 0);
	if (!inst->policies) {
		policy_detach(inst);
		return -1;
	}

	/*
	 *	Parse the policy from the file.
	 */
	if (!rlm_policy_parse(inst->policies, inst->filename)) {
		policy_detach(inst);
		return -1;
	}

	*instance = inst;

	return 0;
}


/*
 *	Insert a named policy into a list.
 */
int rlm_policy_insert(rbtree_t *head, policy_named_t *policy)
{
	if (!rbtree_insert(head, policy)) {
		return 0;
	}

	return 1;
}


/*
 *	Find a named policy
 */
policy_named_t *rlm_policy_find(rbtree_t *head, const char *name)
{
	policy_named_t mypolicy;

	mypolicy.name = name;

	return rbtree_finddata(head, &mypolicy);
}


/*
 *	Find the named user in this modules database.  Create the set
 *	of attribute-value pairs to check and reply with for this user
 *	from the database. The authentication code only needs to check
 *	the password, the rest is done here.
 */
static int policy_authorize(void *instance, REQUEST *request)
{
	return rlm_policy_evaluate((rlm_policy_t *) instance, request,
				   "authorize");
}


static int policy_preacct(void *instance, REQUEST *request)
{
	return rlm_policy_evaluate((rlm_policy_t *) instance, request,
				   "preacct");
}

static int policy_accounting(void *instance, REQUEST *request)
{
	return rlm_policy_evaluate((rlm_policy_t *) instance, request,
				   "accounting");
}

static int policy_post_auth(void *instance, REQUEST *request)
{
	return rlm_policy_evaluate((rlm_policy_t *) instance, request,
				   "post-auth");
}

static int policy_pre_proxy(void *instance, REQUEST *request)
{
	return rlm_policy_evaluate((rlm_policy_t *) instance, request,
				   "pre-proxy");
}

static int policy_post_proxy(void *instance, REQUEST *request)
{
	return rlm_policy_evaluate((rlm_policy_t *) instance, request,
				   "post-proxy");
}

/*
 *	The "free" functions are here, for no particular reason.
 */
void rlm_policy_free_item(policy_item_t *item)
{
	while (item) {
		policy_item_t *next = item->next;

		switch (item->type) {
		default:
		case POLICY_TYPE_BAD:
			break;

		case POLICY_TYPE_ASSIGNMENT:
			{
				policy_assignment_t *this;

				this = (policy_assignment_t *) item;
				if (this->lhs) free(this->lhs);
				if (this->rhs) free(this->rhs);
			}
			break;

		case POLICY_TYPE_CONDITIONAL:
			{
				policy_condition_t *this;

				this = (policy_condition_t *) item;
				if (this->lhs) free(this->lhs);
				if (this->rhs) free(this->rhs);

				if (this->child) {
					rlm_policy_free_item(this->child);
					this->child = NULL;
				}
			}
			break;

		case POLICY_TYPE_IF:
			{
				policy_if_t *this;

				this = (policy_if_t *) item;
				if (this->condition) {
					rlm_policy_free_item(this->condition);
					this->condition = NULL;
				}
				if (this->if_true) {
					rlm_policy_free_item(this->if_true);
					this->if_true = NULL;
				}
				if (this->if_false) {
					rlm_policy_free_item(this->if_false);
					this->if_false = NULL;
				}
			}
			break;

		case POLICY_TYPE_ATTRIBUTE_LIST:
			{
				policy_attributes_t *this;

				this = (policy_attributes_t *) item;
				rlm_policy_free_item(this->attributes);
			}
			break;

		case POLICY_TYPE_NAMED_POLICY:
			{
				policy_named_t *this;

				this = (policy_named_t *) item;
				rad_assert(this->name != NULL);
				free(this->name);
				rlm_policy_free_item(this->policy);
			}
			break;

		case POLICY_TYPE_CALL:
			{
				policy_call_t *this;

				this = (policy_call_t *) item;
				if (this->name) free(this->name);
			}
			break;

		case POLICY_TYPE_RETURN:
			break;	/* do nothing */

		case POLICY_TYPE_MODULE:
			{
				policy_module_t *this;

				this = (policy_module_t *) item;
				if (this->cs) cf_section_free(&this->cs);
				if (this->mc) modcallable_free(&this->mc);
			}
			break;
		} /* switch over type */
		item->next = NULL; /* for debugging & sanity checks */
		item->type = POLICY_TYPE_BAD;
		free(item);

		item = next;
	}
}


/*
 *	The module name should be the only globally exported symbol.
 *	That is, everything else should be 'static'.
 *
 *	If the module needs to temporarily modify it's instantiation
 *	data, the type should be changed to RLM_TYPE_THREAD_UNSAFE.
 *	The server will then take care of ensuring that the module
 *	is single-threaded.
 */
module_t rlm_policy = {
	RLM_MODULE_INIT,
	"policy",
	RLM_TYPE_CHECK_CONFIG_SAFE | RLM_TYPE_HUP_SAFE,   	/* type */
	policy_instantiate,		/* instantiation */
	policy_detach,			/* detach */
	{
		NULL,			/* authentication */
		policy_authorize,	/* authorization */
		policy_preacct,		/* preaccounting */
		policy_accounting,	/* accounting */
		NULL,			/* checksimul */
		policy_pre_proxy,	/* pre-proxy */
		policy_post_proxy,	/* post-proxy */
		policy_post_auth	/* post-auth */
	},
};
